### PR TITLE
New feature: Remote download on content pages

### PR DESCRIPTION
**Pull request by MisRob**
## Summary

- Fix: a remote content page now loads
- Fix: a new download request is now saved where it was meant to in `useDownloadRequests`
- New feature: Adds frontend part of download logic to content pages when browsing remote content. Note that the download action is mocked (see follow-ups and code comments)

| (1) All actions as icons | (2) Actions hidden in the menu | (3) Actions partially hidden in the menu |
| ----- | ----- | ----- |
| Peek 2023-04-27 13-32 | mobile | mobile2 |

Notice somewhat strange UX in (3) as the download option disappears and later the bookmark option appears. We may need to re-think the logic around displaying icons/menu items in the `LearningActivityBar` to be able to improve this. Unless someone has a better idea, I can open an issue for that.

## References

- Resolves a large portion of https://github.com/learningequality/kolibri/issues/9856 but contains some mocks. https://github.com/learningequality/kolibri/issues/10564 needs to be resolved for it to be complete.

- Related KDS PR: https://github.com/learningequality/kolibri-design-system/pull/433

**Follow-ups and related issues**
- https://github.com/learningequality/kolibri/issues/10564
- https://github.com/learningequality/kolibri/issues/10575
- https://github.com/learningequality/kolibri/issues/10565
- https://github.com/learningequality/kolibri/issues/10576
- https://github.com/learningequality/kolibri/issues/10577

## Reviewer guidance

Taking into account the number of related issues, some of them in need of implementation before this feature is complete, I think it'd make more sense to do a more thorough QA of download as a whole as soon we have everything in place. In this PR, we can focus on:

- [ ] Download/bookmark buttons and loader UX
- [ ] Code review
- [ ] Review of related KDS PR https://github.com/learningequality/kolibri-design-system/pull/433
- [ ] QA of conditions for displaying download and bookmark buttons on remote content page:
  - [ ] Neither bookmark nor download button is displayed when logged out
  - [ ] For content that was imported by an admin, the bookmark button is displayed and the download button is not
  - [ ] For content that was downloaded by a learner, the bookmark button is displayed and the download button is not (download is mocked)
  - [ ] For content that wasn't imported by an admin nor downloaded by a learner yet, the bookmark button is not displayed and the download button is displayed
- [ ] QA of regressions on non-remote content pages
  - [ ] Bookmarking still working?

To test remote content, you can navigate to _Library -> Mr Tibbles' Classroom Server -> Kolibri QA Channel_ as a learner

## Comments

- After the referenced KDS PR is merged and before this PR is merged, I will replace the package.json reference
- The original draft PR contains [a comment with the summary of some decisions](https://github.com/learningequality/kolibri/pull/10285#issuecomment-1486716508) that I implemented here
   
----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js [hunk 1]
```diff
@@ -119,11 +119,18 @@ export default function useDownloadRequests(store) {
       status: 'QUEUED',
       date_added: new Date(),
     };
+    // TODO: Remove and replace by real progress implementation
```
Comment (MisRob) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js:122]: Follow-up issue https://github.com/learningequality/kolibri/issues/10564

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js [hunk 2]
```diff
@@ -146,12 +153,48 @@ export default function useDownloadRequests(store) {
...
+
+  // Note that backend filters out download requests that correspond to content that was
+  // removed after being downloaded, therefore the presence of a completed download request
+  // means that the content is downloaded on the device.
+  function isDownloadedByLearner(content) {
+    if (!content || !content.id) {
+      return false;
+    }
+    const downloadRequest = downloadRequestMap.downloads[this.content.id];
+    // TODO: Get real progress from `downloadRequest` as soon as backend is ready
```
Comment (MisRob) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js:179]: Follow-up issue https://github.com/learningequality/kolibri/issues/10564

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js [hunk 3]
```diff
@@ -146,12 +153,48 @@ export default function useDownloadRequests(store) {
     return Promise.resolve();
   }
 
+  function isDownloadingByLearner(content) {
+    if (!content || !content.id) {
+      return false;
+    }
+    const downloadRequest = downloadRequestMap.downloads[this.content.id];
+    // TODO: Get real progress from `downloadRequest` as soon as backend is ready
```
Comment (MisRob) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js:161]: Follow-up issue https://github.com/learningequality/kolibri/issues/10564

File: kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue [hunk 4]
```diff
@@ -32,6 +32,25 @@
       </template>
 
       <template #actions>
+        <Transition name="downloading-loader">
+          <span>
+            <KCircularLoader
```
Comment (MisRob) [inline kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:37]: Follow-up (stretch goal) issue https://github.com/learningequality/kolibri/issues/10565

**Comment by MisRob** [comment]
@jtamiace If you could look at the summary with the three recordings in the PR description, that'd be appreciated. Also note three related issues that involve UX

- https://github.com/learningequality/kolibri/issues/10575
- https://github.com/learningequality/kolibri/issues/10576
- https://github.com/learningequality/kolibri/issues/10565

**Comment by jtamiace** [comment]
@MisRob for (1) and (3) the icons shifting around feels a bit unexpected and distracting. For (1), is it possible to make the spinner replace the location of the download icon so that either the loading spinner or the bookmark icon is showing but not both? So something like: 

Screenshot 2023-05-02 at 2 37 03 PM

For how this adapts on mobile displays (2) and (3), I am not sure what technical solution makes sense, but here is an option that would require some different logic than full width displays but could eliminate the icon shifts. Keeping the download status outside of the menu seems like a nice way to communicate the system status:
Screenshot 2023-05-02 at 2 54 00 PM

^Although I guess disabling could also be used for the desktop view?

**Comment by MisRob** [comment]
Thank you, @jtamiace, that's right. I used the "disabled" state for both mobile and desktop as it was a bit more straightforward to implement. Looking much better now:

recording-mobile

recording-desktop

**Comment by MisRob** [comment]
Related https://github.com/learningequality/kolibri-design-system/pull/433 has been merged and KDS reference in this PR updated to point to the latest KDS `develop`

**Comment by LianaHarris360** [comment]
@MisRob For some resources the KCircularLoader seems to be continuous and is displayed before I have hit download. On other resources, it functions as expected.

https://user-images.githubusercontent.com/46411498/236940656-7ee8bf50-44d5-48c0-8ae5-654b320e0431.mov

**Comment by MisRob** [comment]
@LianaHarris360 Ah, I should have mentioned this - resources with ids `2ea9bda8703241be89b5b9fd87f88815` and `9e53d545aaf44c3787a29a34b189c56a` are currently [hardcoded in `useDownloadRequests`](https://github.com/learningequality/kolibri/blob/d587c87b9baa697c3b5eef4bb83091bf18ebe666/kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js#L37-L66) so if you're seeing this behavior only for these two, that should be fine for now, because they're indefinitely being downloaded due to that mock. Did you observe it for any other resources (you could have a look in URL where a resource ID can be found and compare to those I mentioned)?

**Comment by LianaHarris360** [comment]
Ok, this behavior does only happen for those two hardcoded resources, which makes complete sense. Thank you!

File: kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js [hunk 5]
```diff
@@ -41,6 +41,8 @@ export function showTopicsContent(store, id, deviceId = null) {
   if (deviceId) {
     return setCurrentDevice(deviceId).then(device => {
       const baseurl = device.base_url;
+      const { fetchUserDownloadRequests } = useDownloadRequests(store);
+      fetchUserDownloadRequests({ page: 1, pageSize: 100 });
```
Comment (marcellamaki) [inline kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js:45]: Non-blocking, and can definitely be adjusted in follow up as needed, but do we want the default page size to be 100 here? That seems like a lot to me and like we want smaller pagination segments.

File: kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue [hunk 6]
```diff
@@ -32,6 +32,29 @@
       </template>
 
       <template #actions>
+        <Transition name="downloading-loader">
+          <!--
+            wrapping span needed here because
+            (1) tooltip doesn't display when `ref` is on `KCircularLoader`
```
Comment (marcellamaki) [inline kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:38]: > 1) tooltip doesn't display when `ref` is on `KCircularLoader

Today I learned!

File: kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue [hunk 7]
```diff
@@ -430,16 +497,16 @@
...
           ) {
             this.closeMenu({ focusMoreOptionsButton: false });
           }
         }
         if (this.isTimerOpen) {
           if (
-            !this.$refs.timerButton.$el.contains(event.target) &&
-            !this.$refs.timer.$el.contains(event.target)
+            !this.$refs.timerButton?.$el.contains(event.target) &&
+            !this.$refs.timer?.$el.contains(event.target)
```
Comment (marcellamaki) [inline kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:509]: I am not clear on what this change, and the one above on line 500 and 501 is doing

**Review by marcellamaki — APPROVED** [review]
Thanks @MisRob - I've added just a couple of non-blocking comments from the code read through, but all can be discussed/addressed later. I also think it makes sense to have more robust manual QA happen here _after_ I finish the non-mocked version of the content download request viewset work, so I think we're good to go with @LianaHarris360's manual QA and the QA team can pick up the testing once more of the backend is connected up (cc @radinamatic)

File: kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js (hunk 5 above)
Comment (MisRob) [reply kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js:45]: That's a good point. I don't know if we'll use pagination parameters here on this page, but it's currently how `fetchUserDownloadRequests` is implemented so better to be ready for it. I will resolve it in a follow-up PR.

File: kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue (hunk 6 above)
Comment (MisRob) [reply kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:38]: I don't know if it's the case generally, it's possible that the new `show` prop I added to `KCircularLoader` plays a role in this but I didn't find time to investigate more closely. It's near the end of my TODOs list though :)

File: kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue (hunk 7 above)
Comment (MisRob) [reply kolibri/plugins/learn/assets/src/views/LearningActivityBar.vue:509]: That comes from this commit https://github.com/learningequality/kolibri/pull/10558/commits/5ae63b50a7b22a243db8219dd15a34e8979525a8. At some point, I noticed "Cannot read properties of undefined" console errors and found out it was caused by references not being present at the time of attempting to access them here in some situations. Most likely it was a little bug not directly related to the new code. I added safeguards to resolve that. If you were asking about the syntax, that's [Optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining).

File: kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js (hunk 5 above)
Comment (MisRob) [reply kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js:45]: Follow-up PR https://github.com/learningequality/kolibri/pull/10683

**Pull request merged**